NIFI-10797 add customisable Elasticsearch REST Client config and Elasticsearch Cluster Sniffer

ChrisSamo632 opened the pull request:
# Summary

[NIFI-10797](https://issues.apache.org/jira/browse/NIFI-10797) add customisable Elasticsearch REST Client config and Elasticsearch Cluster Sniffer

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [x] JDK 11
  - [x] JDK 17

### Licensing

- ~[ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)~
- ~[ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files~

### Documentation

- ~[ ] Documentation formatting appears as expected in rendered files~

gresockj commented:
Neat ideas for additional ES client configuration.  Reviewing now

gresockj reviewed (changes requested):
Very cool.

<img width="354" alt="image" src="https://user-images.githubusercontent.com/423565/217235619-02f48ed7-aeae-4866-96be-12462ec739ae.png">

This felt natural to use in NiFi -- it looks like a useful set of additional configuration for the ES client service!

However, I did run into problems using the Sniffer.  With node sniffing off, a SearchElasticsearch query worked, but when I turned node sniffing on in the client, I got the following exception:
```2023-02-07 06:56:00,075 ERROR [es_rest_client_sniffer[T#1]] org.elasticsearch.client.sniff.Sniffer error while sniffing nodes
java.net.ConnectException: Timeout connecting to [/172.17.0.3:9200]
	at org.elasticsearch.client.RestClient.extractAndWrapCause(RestClient.java:881)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:283)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:270)
	at org.elasticsearch.client.sniff.ElasticsearchNodesSniffer.sniff(ElasticsearchNodesSniffer.java:105)
	at org.elasticsearch.client.sniff.Sniffer.sniff(Sniffer.java:208)
	at org.elasticsearch.client.sniff.Sniffer$Task.run(Sniffer.java:138)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Timeout connecting to [/172.17.0.3:9200]
	at org.apache.http.nio.pool.RouteSpecificPool.timeout(RouteSpecificPool.java:169)
	at org.apache.http.nio.pool.AbstractNIOConnPool.requestTimeout(AbstractNIOConnPool.java:628)
	at org.apache.http.nio.pool.AbstractNIOConnPool$InternalSessionRequestCallback.timeout(AbstractNIOConnPool.java:894)
	at org.apache.http.impl.nio.reactor.SessionRequestImpl.timeout(SessionRequestImpl.java:184)
	at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processTimeouts(DefaultConnectingIOReactor.java:214)
	at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvents(DefaultConnectingIOReactor.java:158)
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor.execute(AbstractMultiworkerIOReactor.java:351)
	at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.execute(PoolingNHttpClientConnectionManager.java:221)
	at org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1.run(CloseableHttpAsyncClientBase.java:64)
	... 1 common frames omitted
```

If I added `Sniff on Failure = true`, I then got the following exception:

```
org.apache.nifi.elasticsearch.ElasticsearchException: org.apache.nifi.elasticsearch.ElasticsearchException: java.lang.IllegalStateException: sniffer was not set, unable to sniff on failure
	at org.apache.nifi.elasticsearch.ElasticSearchClientServiceImpl.search(ElasticSearchClientServiceImpl.java:793)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.nifi.controller.service.StandardControllerServiceInvocationHandler.invoke(StandardControllerServiceInvocationHandler.java:254)
	at org.apache.nifi.controller.service.StandardControllerServiceInvocationHandler.invoke(StandardControllerServiceInvocationHandler.java:105)
	at com.sun.proxy.$Proxy160.search(Unknown Source)
	at org.apache.nifi.processors.elasticsearch.AbstractPaginatedJsonQueryElasticsearch.doQuery(AbstractPaginatedJsonQueryElasticsearch.java:128)
	at org.apache.nifi.processors.elasticsearch.AbstractPaginatedJsonQueryElasticsearch.doQuery(AbstractPaginatedJsonQueryElasticsearch.java:46)
	at org.apache.nifi.processors.elasticsearch.AbstractJsonQueryElasticsearch.onTrigger(AbstractJsonQueryElasticsearch.java:213)
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27)
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1356)
	at org.apache.nifi.controller.tasks.ConnectableTask.invoke(ConnectableTask.java:246)
	at org.apache.nifi.controller.scheduling.AbstractTimeBasedSchedulingAgent.lambda$doScheduleOnce$0(AbstractTimeBasedSchedulingAgent.java:59)
	at org.apache.nifi.engine.FlowEngine$2.run(FlowEngine.java:110)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.nifi.elasticsearch.ElasticsearchException: java.lang.IllegalStateException: sniffer was not set, unable to sniff on failure
	at org.apache.nifi.elasticsearch.ElasticSearchClientServiceImpl.runQuery(ElasticSearchClientServiceImpl.java:525)
	at org.apache.nifi.elasticsearch.ElasticSearchClientServiceImpl.search(ElasticSearchClientServiceImpl.java:790)
	... 22 common frames omitted
Caused by: java.lang.IllegalStateException: sniffer was not set, unable to sniff on failure
	at org.elasticsearch.client.sniff.SniffOnFailureListener.onFailure(SniffOnFailureListener.java:59)
	at org.elasticsearch.client.RestClient.onFailure(RestClient.java:526)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:282)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:270)
	at org.apache.nifi.elasticsearch.ElasticSearchClientServiceImpl.performRequest(ElasticSearchClientServiceImpl.java:955)
	at org.apache.nifi.elasticsearch.ElasticSearchClientServiceImpl.runQuery(ElasticSearchClientServiceImpl.java:523)
	... 23 common frames omitted
```

Do I have to do anything to the ES server setup to enable sniffing?  I'm running ES 8.2.0.

Once this is resolved, I didn't have any other concerns with the code.

ChrisSamo632 commented:
@gresockj thanks for the feedback

The "Sniff on Failure" error looks to be due to a bug in the new code - I'd not added the `Sniffer` into the `SniffOnFailureListener`, so I've corrected that.

Are you connecting to Elasticsearch over `http` or `https` (8.x comes with TLS out of the box, but can be turned off)?

I've tried to cater for both protocols in this code, but quite possible there's a problem there. The Sniffer only returns the hostname and port number (no protocol) so we have to tell it whether to use `http` or `https` when generating the host URL - for this, I'm using the "first" of the specified `HTTP_HOSTS` and checking whether it starts with `https://` (otherwise assuming it's `http://`) - I'll add a note to the `Sniffer` property to indicate that all `HTTP_HOSTS` must use the same protocol (which I suspect NiFi and probably Elasticsearch would struggle with anyway if they didn't).

Looking at the IP in your error message, I'm guessing you're running Elasticsearch as a Docker container, is NiFi running in the same Docker network (although if NiFi can connect to the same Elasticsearch instance when the Sniffer is **not** configured, then I suspect it's not a networking issue)?

The Sniffer uses the `Nodes Info` API in Elasticsearch, the only [pre-requisite](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-info.html#cluster-nodes-info-api-prereqs) for use is that if Elasticsearch Security is enabled (default to `true` in 8.x) then the user connecting must have the `monitor` or `manage` [cluster privilege](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-privileges.html#privileges-list-cluster) to successfully list the cluster nodes - I'll note this in the `Sniffer` property too. But I'd expect that to throw a `403` (or similar) error response rather than causing a connection timeout.

ChrisSamo632 commented:
The `integration-tests` (with update `verifySniffer` logic) just revealed that the Elasticsearch instance's `network.publish_address` **must** be one of the `network.bind_address`(es) for the Sniffer to work - this wasn't the case by default for the TestContainer running Elasticsearch, so it could be an issue like that - again, I've added a note to the `Sniffer` property and the `integration-tests` now set `network.bind_address` to [site](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#network-interface-values), which allows them to pass on both Elasticsearch 7.x and 8.x (no TLS but with Auth enabled)

Going back to the earlier Docker question - is it that you're running Elasticsearch as a Docker container but NiFi "natively" on your machine, connecting to Elasticsearch using `localhost` and then the Elasticsearch instance is reporting back the Docker-internal address of `172.17.0.3`? What do you see under the `http` section of the response from a call to your Elasticsearch's `GET /_nodes/http` API?

With an Elasticsearch Docker container running, I get:
```json
"http": {
  "bound_address": [
    "0.0.0.0:9200"
  ],
  "publish_address": "172.20.0.2:9200",
  "max_content_length_in_bytes": 104857600
}
```
which isn't contactable from my browser/terminal - so I suspect this is indeed what's going on for you, i.e. your Elasticsearch instance is reporting back its "internal" IP as its `publish_address` and you'd need to re-configure the instance's `network.publish_address` (or run NiFi within the Docker network **or** use the `host` network for your Elasticsearch container) so that the correct combination of `publish_address` and host network routing can be used

ChrisSamo632 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -220,16 +285,36 @@ public List<ConfigVerificationResult> verify(final ConfigurationContext context,
+    private void verifySniffer(final ConfigurationContext context, final RestClient verifyClient, final ConfigVerificationResult.Builder snifferResult) {
+        try (final Sniffer verifySniffer = setupSniffer(context, verifyClient)) {
+            if (verifySniffer != null && elasticsearchNodesSniffer != null) {
+                final List<Node> nodes = elasticsearchNodesSniffer.sniff();
```
Now that we get a definite set of Nodes through the Sniffer during verification, we could take the opportunity to confirm connectivity - as noted in the PR conversations, it can be that the Elasticsearch instance `publish_address` is an "internal" area that can't actually be contacted by NiFi

While we can't change how the Sniffer works here, we can provide helpful guidance for NiFi users using the verification checks, e.g. "Found 1 node, 0 nodes were contactable, check the Elasticsearch Network Settings". This wouldn't prevent connectivity issues when using the Sniffer for real (it would still get the publish_address from Elasticsearch), but should help NiFi users understand what needs changing in Elasticsearch (or whether the Sniffer is the right option in their use case)

Maybe also link to the Elasticsearch docs from this Controller's additionalDetails?

ChrisSamo632 commented:
@gresockj the [Elasticsearch documentation/blogs](https://www.elastic.co/blog/elasticsearch-sniffing-best-practices-what-when-why-how) indicate that this is indeed what's going on (I've added an "additional details" section for the Controller Service and linked to these docs).

I also added a connectivity check to the "Verify" steps when the Sniffer is enabled - the `integration-tests` were updated to cater for this too (see the updated `AbstractElasticsearchITBase` for the settings I changed to make the `TestContainer` contactable after Sniffing as well as by the standard RestClient setup)

gresockj commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/resources/docs/org.apache.nifi.elasticsearch.ElasticSearchClientServiceImpl/additionalDetails.html:
```diff
@@ -0,0 +1,54 @@
+    Elasticsearch Networking Advanced Settings</a>, such as <code>network.publish_host</code> to ensure that
+    the HTTP Hosts found by the Sniffer are accessible by NiFi. For example, Elasticsearch may use a network internal
+    <code>publish_address</code> that is inaccessible to NiFi, but instead should use an address/IP that NiFi understands.
+    It may also be necessary to add this same address to Elasticsearch's <code>network.bind_address</code> list.
```
`network.bind_host`

gresockj reviewed (changes requested):
Thanks for the comments, which helped me correct the networking configuration issue.  Just one more change needed, as noted below.

gresockj commented:
> Going back to the earlier Docker question - is it that you're running Elasticsearch as a Docker container but NiFi "natively" on your machine, connecting to Elasticsearch using `localhost` and then the Elasticsearch instance is reporting back the Docker-internal address of `172.17.0.3`? What do you see under the `http` section of the response from a call to your Elasticsearch's `GET /_nodes/http` API?

Yes, that is my setup, and I get similar output to what you posted.

> which isn't contactable from my browser/terminal - so I suspect this is indeed what's going on for you, i.e. your Elasticsearch instance is reporting back its "internal" IP as its `publish_address` and you'd need to re-configure the instance's `network.publish_address` (or run NiFi within the Docker network **or** use the `host` network for your Elasticsearch container) so that the correct combination of `publish_address` and host network routing can be used

Good catch -- when I added `-e network.publish_host=192.168.1.X` to my docker run command, and used the same IP in the NiFi ES client service, the sniffer worked correctly.

ChrisSamo632 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service-api/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientService.java:
```diff
@@ -161,6 +157,118 @@ public interface ElasticSearchClientService extends ControllerService, Verifiabl
+                    "have the \"monitor\" or \"manage\" cluster privilege to use this API." +
+                    "Note that all " + HTTP_HOSTS.getDisplayName() + " (and those that may be discovered within the cluster " +
+                    "using the Sniffer) must use the same protocol, e.g. http or https, and be contactable using the same client settings. " +
+                    "Finally the Elasticsearch \"network.publish_address\" must match one of the \"network.bind_address(es)\" " +
```
Darn, for some reason I keep typing `_address`, good catch though, think I\ve updated it everywhere for both `bind_` and `publish_` settings
